src/e3/testsuite/control.py

```diff
@@ -0,0 +1,251 @@
+            def error(message):
+                raise ValueError("entry #{}: {}".format(i, message))
+
+            if (
+                not isinstance(entry, list) or
+                not len(entry) in (2, 3) or
+                any(not isinstance(s, str) for s in entry)
+            ):
+                error("list of 2 or 3 strings expected")
+
+            # Decode the test control kind
+            try:
+                kind = TestControlKind[entry[0]]
+            except KeyError:
+                error("invalid kind: {}".format(entry[0]))
+
+            # Evaluate the condition
+            try:
+                cond = eval(entry[1], condition_env)
+            except Exception as exc:
+                error("invalid condition ({}): {}"
+                      .format(type(exc).__name__, exc))
+
+            message = entry[2] if len(entry) > 2 else None
+
+            entries.append((kind, cond, message))
+
+        # Now, select the first entry whose condition is True. By default,
+        # fallback to "default".
+        for kind, cond, message in entries:
+            if cond:
+                skip, xfail = {
+                    TestControlKind.NONE: (False, False),
+                    TestControlKind.SKIP: (True, False),
+                    TestControlKind.XFAIL: (False, True),
+                }[kind]
+                return TestControl(message, skip, xfail)
+        return default
+
+
+class OptfileTestControlCreator(TestControlCreator):
+    """Create test controls based on "test.opt" files."""
+
+    def default_script(self, driver):
+        """Return the default test script filename. """
+        # Use "test.cmd" by default. If it does not exist while there is a
+        # "test.py" file, use that instead.
+        if (
+            not os.path.isfile(driver.test_dir("test.cmd"))
+            and os.path.isfile(driver.test_dir("test.py"))
+        ):
+            return "test.py"
+        return "test.cmd"
+
+    def default_opt_results(self, driver):
+        """
+        Return the default options. test.opt files can override these.
+
+        By default, a test is not DEAD, SKIP, nor XFAIL. Its execution timeout
+        is 780 seconds. Test script is "test.cmd" and its output is compared
+        against "test.out".
+        """
+        return {
+            "RLIMIT": "780",
+            "DEAD": None,
+            "XFAIL": None,
+            "SKIP": None,
+            "OUT": "test.out",
+            "CMD": self.default_script(driver),
+            "FILESIZE_LIMIT": None,
+            "TIMING": None,
+            "NOTE": None,
+        }
+
+    def __init__(self, system_tags, opt_filename="test.opt"):
+        """
+        :param str|list[str] system_tags: Tags to forward to OptFileParse().
+        :param str opt_filename: Name of the file to parse, relative to the
+            test directory.
+        """
+        self.opt_filename = opt_filename
+        self.system_tags = system_tags
+
+    def create(self, driver):
```

Comment:
docstring

Reply:
No need: this is an overriding method which does exactly what the base method documents